File: src/main/java/com/senderman/lastkatkabot/DBService.java
package com.senderman.lastkatkabot;

import com.senderman.lastkatkabot.tempobjects.BnCPlayer;
import com.senderman.lastkatkabot.tempobjects.BullsAndCowsGame;
import org.telegram.telegrambots.meta.api.objects.Message;

import java.util.List;
import java.util.Map;
import java.util.Set;

public interface DBService {

    void initStats(int id);

    void incTotalDuels(int id);

    void incDuelWins(int id);

    void incBNCWins(int id, int points);

    Map<String, Integer> getStats(int id);

    List<BnCPlayer> getTop(); // return map of <id, score> sorted by descending order

    void setUserCity(int id, String city);

    String getUserCity(int id);


    void addTgUser(int id, COLLECTION_TYPE type);

    void removeTGUser(int id, COLLECTION_TYPE type);

    Set<Integer> getTgUsersFromList(COLLECTION_TYPE type);

    Set<Integer> getTgUsersIds(COLLECTION_TYPE type);

    Set<Integer> getAllUsersIds();


    void addUserToChatDB(Message message);

    void removeUserFromChatDB(int userId, long chatId);

    List<Integer> getChatMemebersIds(long chatId);

    void removeOldUsers(long chatId, int date);


    Map<Long, BullsAndCowsGame> getBnCGames();

    void saveBncGame(long chatId, BullsAndCowsGame game);

    void deleteBncGame(long chatId);


    int getTournamentMessageId();

    void setTournamentMessage(int messageId);


    void addAllowedChat(long chatId, String title);

    Set<Long> getAllowedChatsSet();

    void updateChatId(long oldChatId, long newChatId);

    void updateTitle(long chatId, String title);

    void removeAllowedChat(long chatId);

    void setPair(long chatId, String pair);

    Map<Long, String> getAllowedChats();

    void cleanup();


    boolean pairExistsToday(long chatId);

    enum COLLECTION_TYPE {ADMINS, PREMIUM, BLACKLIST}

    String getPairOfTheDay(long chatId);

    String getPairsHistory(long chatId);


    // will be removed after a while

    void incRavenMessages(int date);

    int getRavenMessages();

    void updateRavenRecord();

    void incInterruptions();

    void redInterruptions();

    int getInterruptions();

    int getLastRavenDate();

    int getRavenRecord();

}

File: src/main/java/com/senderman/lastkatkabot/DBService.kt
package com.senderman.lastkatkabot

import com.senderman.lastkatkabot.bnc.BullsAndCowsGame
import com.senderman.lastkatkabot.tempobjects.UserRow
import com.senderman.lastkatkabot.tempobjects.UserStats
import org.bson.Document
import org.telegram.telegrambots.meta.api.objects.Message

interface DBService {

    enum class UserType {
        ADMINS, PREMIUM, BLACKLIST
    }

    fun initStats(id: Int): Document
    fun incTotalDuels(id: Int)
    fun incDuelWins(id: Int)
    fun incBNCWins(id: Int, points: Int)
    fun getStats(id: Int): UserStats
    // return map of <id, score> sorted by descending order
    fun getTop(): Map<Int, Int>

    fun transferStats(fromId: Int, toId: Int)

    fun setUserCity(id: Int, city: String)
    fun getUserCity(id: Int): String?
    fun setLover(userId: Int, loverId: Int)
    fun getLover(userId: Int): Int
    fun divorce(userId: Int)


    fun addTgUser(id: Int, type: UserType)
    fun removeTGUser(id: Int, type: UserType)
    fun getTgUsersByType(type: UserType): MutableSet<Int>


    fun getAllUsersIds(): Set<Int>
    fun addUserToChatDB(message: Message)
    fun removeUserFromChatDB(userId: Int, chatId: Long)
    fun getChatMemebersIds(chatId: Long): MutableList<Int>
    fun removeOldUsers(chatId: Long, date: Int)


    fun getBnCGames(): MutableMap<Long, BullsAndCowsGame>
    fun saveBncGame(chatId: Long, game: BullsAndCowsGame)
    fun deleteBncGame(chatId: Long)

    fun saveRow(chatId: Long, row: UserRow)
    fun getUserRows(): MutableMap<Long, UserRow>

    fun getTournamentMessageId(): Int
    fun setTournamentMessage(messageId: Int)


    fun addChat(chatId: Long, title: String)
    fun getChatTitleMap(): Map<Long, String>
    fun getChatIdsSet(): MutableSet<Long>
    fun updateChatId(oldChatId: Long, newChatId: Long)
    fun updateTitle(chatId: Long, title: String)
    fun removeChat(chatId: Long)
    fun cleanup()


    fun setPair(chatId: Long, pair: String)
    fun pairExistsToday(chatId: Long): Boolean
    fun getPairOfTheDay(chatId: Long): String
    fun getPairsHistory(chatId: Long): String?
}